src/tools/crawler/rpc.rs

```diff
@@ -37,18 +37,26 @@ pub async fn initialize_rpc_server(rpc_addr: SocketAddr, rpc_context: RpcContext
 fn create_rpc_module(rpc_context: RpcContext) -> RpcModule<RpcContext> {
     let mut module = RpcModule::new(rpc_context);
     module
-        .register_method("getmetrics", |params, rpc_context| {
+        .register_method("getmetrics", |_params, rpc_context| {
+            Ok(rpc_context.0.lock().unwrap().clone())
+        })
+        .unwrap();
+    module
+        .register_method("dumpmetrics", |params, rpc_context| {
             let report_params = params.parse::<ReportParams>()?;
             if let Some(path) = report_params.file {
                 let content = serde_json::to_string(rpc_context.0.lock().unwrap().deref())?;
-                let length = content.len();
+                // Wrap our NetworkSummary in a JSON-RPC response envelope
+                let response =
+                    "{\"jsonrpc\":\"2.0\",\"result\":".to_owned() + &content + ",\"id\":0}";
+                let length = response.len() as i32;
                 // TODO: consider some checks against directory traversal
-                if let Err(e) = fs::write(path, content) {
+                if let Err(e) = fs::write(path, response) {
                     warn!("Unable to write to file: {}", e);
                 }
-                Ok(RpcOutput::Length(length))
+                Ok(length)
             } else {
-                Ok(RpcOutput::Summary(rpc_context.0.lock().unwrap().clone()))
+                Ok(-1)
```

Comment:
Just a little comment could be useful what does would mean to the caller - is this an error code with some meaning or just some random negative to indicate there was an error (path was not found)?

Reply:
yes, this was completely naked here -- I've just checked in a change where an object get returned in the result: with status (success/fail), length, and message.

Reply:
// expected response
```
{
    "jsonrpc": "2.0",
    "result": {
        "status": "success",
        "length": 12400,
        "message": ""
    },
    "id": 0
}
```
// no file parameter
```
{
    "jsonrpc": "2.0",
    "result": {
        "status": "fail",
        "length": -1,
        "message": "No file parameter in params"
    },
    "id": 0
}
```
// bad file path passed in, or other file I/O error
```
{
    "jsonrpc": "2.0",
    "result": {
        "status": "fail",
        "length": -1,
        "message": "Unable to write file: No such file or directory (os error 2)"
    },
    "id": 0
}
```